### PR TITLE
Add metrics to VPN

### PR DESCRIPTION
**Pull request by mcleinman**
## Description

Adding metrics, main code for [Super Dooper Metrics epic](https://mozilla-hub.atlassian.net/browse/VPN-3285). 

## Reference

https://mozilla-hub.atlassian.net/browse/VPN-3630

## Checklist
    
- [X] My code follows the style guidelines for this project
- [X] I have not added any packages that contain high risk or unknown licenses (GPL,  LGPL, MPL, etc. consult with DevOps if in question)
- [X] I have performed a self review of my own code
- [X] I have commented my code PARTICULARLY in hard to understand areas
- [X] I have added thorough tests where needed

### CODE REVIEW

File: src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp
Comment (bakulf) [inline src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp:57]: what about errors? Do we care?

File: src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml
Comment (bakulf) [inline src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml:33]: This is bad. This switch-stmt does not scale. What about if:
1. we make navigator flags a bitmap
2. we introduce a record-to-glean flag
3. in Navigator.cpp you check if the flag contains the record-to-glean, and you call `VPNGleanDeprecated...`.

In this way, if tomorrow we want to record screenFooBar, we can simply pass a flag.

**Review by bakulf — CHANGES_REQUESTED** [review]

**Comment by mcleinman** [comment]
# Request for data collection review form

**All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**

1) What questions will you answer with this data?

These are a handful of metrics to learn more about how people use the VPN app. Which features are not being used? Is it because they're undiscoverable, or other reasons? How long are users keeping a VPN session active for? What sort of account creation and VPN connectivity issues do users face, and how often? 

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements? 

We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? What features are being used, and which aren't? Knowing where the problems are and which features are not used will help us better design a smooth user experience. 

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Carefully considered which metrics were needed.

4) Can current instrumentation answer these questions?

Some of these questions, yes. This work includes better surfacing some existing metrics, and this request is for additional ones.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.   

**Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**

<table>
  <tr>
    <td>Measurement Name</td>
    <td>Measurement Description</td>
    <td>Data Collection Category</td>
    <td>Tracking Bug #</td>
  </tr>
  <tr>
    <td>authentication_error - reason: IncorrectPassword | InvalidEmail | InvalidOrExpiredVerificiationCode | InvalidTOTP | InvalidUnblockCode | Timeout
    </td>
    <td>Registration/login errors</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>user_changed_endpoint_geo</td>
    <td>A user changed their VPN node</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>speed_test_started, speed_test_completed (w/ extra `speed = Slow | Medium | Fast`)</td>
    <td>The start and end of a user-initiated speed test, along with the rough category the speed teest fell into</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>bottom_navigation_bar_click (w/ extra `bar_button = ScreenMessaging | ScreenHome | ScreenSettings | Unknown`)</td>
    <td>Keeping track of user tab clicks in the VPN app</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>connect_on_startup_active</td>
    <td>Boolean on user changing the "start VPN on boot" setting</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>using_system_language</td>
    <td>Boolean on whether users are using the app in the OS system language</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>support_case_submitted</td>
    <td>Event when a user submits a support case</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>settings_about_us_opened</td>
    <td>Event when user taps the "about us" menu in Settings</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>

  <tr>
    <td>help_menu_help_center_opened, help_menu_contact_support_opened, help_menu_view_logs_opened</td>
    <td>Tracking usage of various options in the help menu</td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>


  <tr>
    <td> </td>
    <td> </td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>


  <tr>
    <td> </td>
    <td> </td>
    <td>interaction</td>
    <td><a href="https://mozilla-hub.atlassian.net/browse/VPN-3630">VPN-3630</a>{{</td>
  </tr>
</table>

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is documented in the Glean Dictionary at https://dictionary.telemetry.mozilla.org/"

7) How long will this data be collected?  Choose one of the following:

* I want to permanently monitor this data. mcleinman@mozilla.com

8) What populations will you measure?

No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.

9) If this data collection is default on, what is the opt-out mechanism for users?

When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.

10) Please provide a general description of how you will analyze this data.

Dashboards that product managers and others will consult on a regular basis.

11) Where do you intend to share the results of your analysis?

Within the Mozilla VPN team and Security and Privacy Products team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection? 

No

File: src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp
Comment (mcleinman) [reply src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp:57]: Santiago and I (and Santiago and other people) have gone over [his desired metrics](https://mozilla-hub.atlassian.net/wiki/spaces/SECPRV/pages/117309748/Metrics+Collection+Dashboards+PRD) a lot. Error has not come up as something helpful to his future analysis - but if we need it, we could probably extrapolate from `completed - started`, or add a future metric.

File: src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml
Comment (mcleinman) [reply src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml:33]: 100% agreed, it's some horrible code. This PR is not in review shape, apologies that I didn't catch that it automatically tagged code owners, even though I tried to put in on draft immediately. 

I thought from my testing, doing it in `Navigator.cpp` just recorded an integer, which is still not what we were hoping for here - but I can double check.

Comment (mcleinman) [reply src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml:33]: Confirmed:
```
"name": "bottom_navigation_bar_click",
      "timestamp": 2,
      "extra": {
        "bar_button": "10"
      }
```

That said, there has got to be a better way to handle this for C++ enums. I couldn't find it on a quick search before, will do some more research or ask around tomorrow.

Comment (bakulf) [reply src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml:33]: You have to convert them enum to string. See similar glean events. For instance the update steps.

File: glean/metrics.yaml
Comment (strseb) [inline glean/metrics.yaml:1661]: What is this metric trying to answer? Specifically in relation to "speed_test_completed". 

`type:Event`'s are annoying to analyze, and the relation of multiple event's is worse ... or so i have been told by people i'm annoying about relating events for me. 

What do you think about: 
-  have a single metric `speed_test`. 
- Have it a labeled counter: https://mozilla.github.io/glean/book/reference/metrics/labeled_counters.html
- Each state "good", "bad", "failed" , "canceled" ... etc  is a label
We add +1 on the counter of the outcome. 

That way we can calc the sum and and the distribution of each "bucket" including cancelations and failures

**Comment by mcleinman** [comment]
Tagging @travis79  for data-review

**Review by travis79 — APPROVED** [review]
> # Request for data collection review form
> 
> **All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.**
> 
>     1. What questions will you answer with this data?
> 
> 
> These are a handful of metrics to learn more about how people use the VPN app. Which features are not being used? Is it because they're undiscoverable, or other reasons? How long are users keeping a VPN session active for? What sort of account creation and VPN connectivity issues do users face, and how often?
> 
>     2. Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?
> 
> 
> We do not have great insight into some of basic questions about our users: How often are they having issues? What sort of issues are they? What features are being used, and which aren't? Knowing where the problems are and which features are not used will help us better design a smooth user experience.
> 
>     3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
> 
> 
> Carefully considered which metrics were needed.
> 
>     4. Can current instrumentation answer these questions?
> 
> 
> Some of these questions, yes. This work includes better surfacing some existing metrics, and this request is for additional ones.
> 
>     5. List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.
> 
> 
> **Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.**
> Measurement Name 	Measurement Description 	Data Collection Category 	Tracking Bug #
> authentication_error - reason: IncorrectPassword | InvalidEmail | InvalidOrExpiredVerificiationCode | InvalidTOTP | InvalidUnblockCode | Timeout 	Registration/login errors 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> user_changed_endpoint_geo 	A user changed their VPN node 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> speed_test_started, speed_test_completed (w/ extra `speed = Slow | Medium | Fast`) 	The start and end of a user-initiated speed test, along with the rough category the speed teest fell into 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> bottom_navigation_bar_click (w/ extra `bar_button = ScreenMessaging | ScreenHome | ScreenSettings | Unknown`) 	Keeping track of user tab clicks in the VPN app 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> connect_on_startup_active 	Boolean on user changing the "start VPN on boot" setting 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> using_system_language 	Boolean on whether users are using the app in the OS system language 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> support_case_submitted 	Event when a user submits a support case 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> settings_about_us_opened 	Event when user taps the "about us" menu in Settings 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> help_menu_help_center_opened, help_contact_support_opened, help_menu_view_logs_opened 	Tracking usage of various options in the help menu 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> session_start, session_end, daemon_session_start, daemon_session_end 	Tracking time stamps of when a VPN session is started or ended (both in app and overall, to capture difference as the "amount started from system settings" 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> dns_type (value: AdBlock | AntiTracking | AdBlockAndAntiTracking | Custom | Normal) apps_excluded 	Tracking usage of advanced options (to be clear: the "Custom" dns_type is literally the word "custom", it is NOT the IP address that the user set 	interaction 	[VPN-3630](https://mozilla-hub.atlassian.net/browse/VPN-3630)
> 
>     6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
> 
> 
> This collection is documented in the Glean Dictionary at [dictionary.telemetry.mozilla.org](https://dictionary.telemetry.mozilla.org/)"
> 
>     7. How long will this data be collected?  Choose one of the following:
> 
> 
>     * I want to permanently monitor this data. [mcleinman@mozilla.com](mailto:mcleinman@mozilla.com)
> 
> 
>     8. What populations will you measure?
> 
> 
> No filters - all channels, countries, and locales, unless the user has opted out of data collection on that device.
> 
>     9. If this data collection is default on, what is the opt-out mechanism for users?
> 
> 
> When launching the app for the first time, a user is given an option for whether to allow data collection. After this initial set up screen, a user can always toggle data collection permissions in the System Preferences screen.
> 
>     10. Please provide a general description of how you will analyze this data.
> 
> 
> Dashboards that product managers and others will consult on a regular basis.
> 
>     11. Where do you intend to share the results of your analysis?
> 
> 
> Within the Mozilla VPN team and Security and Privacy Products team.
> 
>     12. Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection?
> 
> 
> No

# Data Review

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way?

Yes, through the metrics.yaml file and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/).

2) Is there a control mechanism that allows the user to turn the data collection on and off?

Yes, through the data preferences in the application settings.

3) If the request is for permanent data collection, is there someone who will monitor the data over time?

Permanent collection to be monitored by mcleinman@mozilla.com

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

Category 2, Interaction Data

5) Is the data collection request for default-on or default-off?

Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

No

7) Is the data collection covered by the existing Firefox privacy notice?

Yes

8) Does the data collection use a third-party collection tool?

No

## Result

data-review+

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1661]: I'm going to drop the speed_test_started metric. It's not called for by the PRD, and if we end up wanting additional speed test outcomes (that weren't requested in the PRD), we can add them to the speed_test_completed test (as you point out).

File: src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml
Comment (mcleinman) [reply src/apps/vpn/ui/navigator/navigationBar/VPNBottomNavigationBarButton.qml:33]: Thanks. I don't believe there is a way to do this on the QML side, right? I have it working on the C++ side - thanks for the tips.

(This PR still isn't quite ready for review - I'll make it clear when it's in decent shape.)

**Comment by mcleinman** [comment]
@travis79 I made a slight adjustment to the `user_changed_endpoint_geo`. It's still an event, but it's now reporting an extra string value of `server`, which is one of these 3 values: `singleHopServer`, `multiHopEntryServer`, `multiHopExitServer`. Should I create a new data collection review form, or is that not needed for adding extra values to an already-approved metric? Thanks!

**Comment by travis79** [comment]
> @travis79 I made a slight adjustment to the `user_changed_endpoint_geo`. It's still an event, but it's now reporting an extra string value of `server`, which is one of these 3 values: `singleHopServer`, `multiHopEntryServer`, `multiHopExitServer`. Should I create a new data collection review form, or is that not needed for adding extra values to an already-approved metric? Thanks!

Typically it would be necessary because it's adding to the data we are collecting, but since this hasn't landed yet we should be fine to just amend the current request. I don't see any reason why the content or addition of that extra info would change the review status, so it's still `data-review+` from my perspective.

**Comment by mcleinman** [comment]
At long last, this is off draft and ready for review. More metrics are coming in separate PRs, all part of the [overall epic](https://mozilla-hub.atlassian.net/browse/VPN-3285).

@brizental The test infrastructure only checks the old Glean version. I considered adding in test infrastructure for the new glean world, but that seemed like it should come in a separate PR. Let me know if you want to block this PR on that work; I'd understand.

**Comment by brizental** [comment]
> @brizental The test infrastructure only checks the old Glean version. I considered adding in test infrastructure for the new glean world, but that seemed like it should come in a separate PR. Let me know if you want to block this PR on that work; I'd understand.

Glean testing APIs are available for all metrics on the "new" Glean. This is not infrastructure we implemented, it's just a Glean provided feature. See for example: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/tests/unit/testaddon.cpp#L1117-L1196. The documentation can be found on the metrics page on the Glean book.

File: glean/metrics.yaml
Comment (brizental) [inline glean/metrics.yaml:1686]: How is this metric different from the myriad of authentication error metrics we already have? https://dictionary.telemetry.mozilla.org/apps/mozilla_vpn?page=1&search=authentication

File: glean/metrics.yaml
Comment (brizental) [inline glean/metrics.yaml:1641]: From the PRD I understand this is coming from the "# of times a speed test is performed" request. Correct?

I would suggest adding a metric for `speed_test_started` as well so we can have a full picture of speed test usage, even if it doesn't complete.

File: glean/metrics.yaml
Comment (brizental) [inline glean/metrics.yaml:1749]: nit:
```suggestion
  help_menu_contact_support_opened:
```
For standardization sake.

File: src/apps/vpn/frontend/navigator.cpp
Comment (brizental) [inline src/apps/vpn/frontend/navigator.cpp:520]: I think this comment is unnecessary.

File: tests/qml/tst_GleanEvents.qml
Comment (brizental) [inline tests/qml/tst_GleanEvents.qml:144]: This is just testing Glean, not the VPN. Ideally we would perform whatever action should trigger a metric recording and then check that it resulted in a metric recording. Also this is testing deprecated Glean. See my other comment on how to test the "new" Glean. Note that the testing APIs are also available from JavaScript. When reading the documentation on the Glean book, refer to the "Firefox Desktop" tab for the correct API.

**Review by brizental — CHANGES_REQUESTED** [review]

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1749]: That was my original dream, but when trying that glean gave an error: `'help_menu_contact_support_opened' is too long`

I could pull "menu" from all the others, but then `help_menu_help_center_opened` looks really weird... I'm open to feedback on this, the lack of consistency hurts me inside as well.

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1641]: I had that originally, but pulled it based on [Basti's feedback above](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5702#discussion_r1084343827). 

I'll ask Santiago on the PRD what his hope is for this one.

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1686]: From conversations with Santiago, he wanted to be able to drill into where the failure happened. The extra keys for this metric (Timeout | IncorrectPassword | InvalidEmail | InvalidOrExpiredVerificiationCode | InvalidTOTP | InvalidUnblockCode) are what makes this different. 

Do you think we should consider removing the old one? (If it's currently used for analysis, this would obviously be subpar.) Or adding these keys to it? Or something else?

File: src/apps/vpn/frontend/navigator.cpp
Comment (mcleinman) [reply src/apps/vpn/frontend/navigator.cpp:520]: I like to leave comments in any situation where a future maintainer many years down the line might say "not sure why they did this" and refactor some code into a worse state (in this case, I could see someone moving the metrics recording to `Navigator::requestScreen`, which would log events in many situations we don't want). My hope is to give clear reason as to why this exists the way it does.

Even with that philosophy (which is probably shared by many of us), I'll acknowledge this is a questionable case as to whether we need a comment. I usually err on the side of an extra comment, when unsure. 

If you think the comment should be removed, just say the word and I'll remove it.

File: tests/qml/tst_GleanEvents.qml
Comment (mcleinman) [reply tests/qml/tst_GleanEvents.qml:144]: That makes sense, thanks. I'll essentially turn these into functional tests, may not finish that until tomorrow.

File: glean/metrics.yaml
Comment (brizental) [reply glean/metrics.yaml:1669]: I'm a bit confused. What do you mean by "where the failure happened"? The extra keys are more about _which_ failure happened, no? If so, that is already answered by https://dictionary.telemetry.mozilla.org/apps/mozilla_vpn/metrics/sample_authentication_inapp_error if I understand correctly. See the "error" extra key. If anything, `authentication_inapp_error` even provides more info about the error.

File: glean/metrics.yaml
Comment (brizental) [reply glean/metrics.yaml:1624]: This circles back to, what exactly do we want to answer with this metric? 

Do we want to understand the usage of the speed test? In that case we would need the "started" metric to get a full picture. Things we may want to know: How many users attempt a speed test and leave before it completes? How long does it take for a speed test to complete? How long before a user gives up? What do users do after the speed test? What events precede a speed test attempt? This metric alone does not allow us to answer all of this. 

With the current approach, we can only say "when a speed test is completed, the result is X". I don't see too much value in that. It looks more like a performance metric, and if we were interested in performance, interaction metrics are likely not the best approach either.

File: glean/metrics.yaml
Comment (brizental) [reply glean/metrics.yaml:1732]: Hum, I was afraid of that. Sadly we can't use categories for metrics unless we are confortable having them only on the new Glean. Otherwise, you would be able to have a help_menu category and each on these events inside it. That would be :chefs-kiss.

File: src/apps/vpn/frontend/navigator.cpp
Comment (brizental) [reply src/apps/vpn/frontend/navigator.cpp:502]: The code looks straight forward enough to not require the comment for the future, but no strong opinions.

**Comment by brizental** [comment]
> We wrapped a bunch of old Glean in the helper.cpp file for testing, and I was thinking we'd want to continue that pattern - perhaps tucking the new glean functions in the existing helper functions, so that we're automatically sending metrics to both glean instances when we call those functions.

I would prefer to just straight up use the glean testing API in the testing suites where we need to test Telemetry and I wouldn't worry too much about testing deprecated Glean. If this becomes too much hassle, as much as it pains me to say this, just file a follow up ticket for the work.

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1732]: I'd have a preference towards not requiring new glean, to be able to collect on iOS. I know iOS changes are coming to use new glean, but it's possible/likely that it will be a release or so behind this work.

That said, don't feel strongly about that - let me know if you want me to move this to new glean and add a category.

File: src/apps/vpn/frontend/navigator.cpp
Comment (mcleinman) [reply src/apps/vpn/frontend/navigator.cpp:502]: In that case I'm keeping it, as I don't think it harms.

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1669]: Santiago was most concerned with "what error does the user see", and seemingly less concerned about what is happening technically behind the scenes. (A metric that gives both "what users see" and "what is happening technically" would be best.) I'll acknowledge that I didn't see this existing metric until you pointed it out in your earlier comment, but also - looking at the flow, it doesn't seem like it captures many things that Santiago is interested in - see case 183 and 103 for example. A whole bunch of that switch statement doesn't end up logging something in glean.

I can add add glean logging for all the additional errors not currently covered, then provide a translation somewhere of which error numbers map to which user-shown errors. Over time as code changes are made, that's likely to diverge and not be kept up to date. 

It seems like we have two approaches here:
1) (How it's currently written.) Log technical errors and user-facing errors separately in glean, with different metrics. It's a little duplicative, but it stays up to date with no engineering work, and it's easier to analyze.
2) Log both in one glean metric. Do some post-processing to separate all the technical details into the user-facing errors. (AKA: Data analysts have additional code, rather than our codebase as in the #1.) Try to provide comments in code so that we don't forget to update that technical-error-to-user-error-translation guide when code changes in future years, but it's higher risk that the user-facing error metric will fall out of date at some point.

I think I have a slight inclination towards #1, as it's easier for analysis and more future-proof. (Part of this project is about making it easier for PMs and others to run ad-hoc queries.) I definitely don't feel strongly about this though - despite the length of what I just wrote (sorry) - and so if our approach is usually closer to number 2, let me know and I'll implement it.

**Comment by mcleinman** [comment]
> > We wrapped a bunch of old Glean in the helper.cpp file for testing, and I was thinking we'd want to continue that pattern - perhaps tucking the new glean functions in the existing helper functions, so that we're automatically sending metrics to both glean instances when we call those functions.
> 
> I would prefer to just straight up use the glean testing API in the testing suites where we need to test Telemetry and I wouldn't worry too much about testing deprecated Glean. If this becomes too much hassle, as much as it pains me to say this, just file a follow up ticket for the work :(

I'm not sure we're able to use our unit testing to do much more than is implemented in the qml tests, unfortunately. I think "does it log when we tap something" probably belongs in functional tests, which isn't possible with our current setup. Ticket added: https://mozilla-hub.atlassian.net/browse/VPN-4149

File: glean/metrics.yaml
Comment (mcleinman) [reply glean/metrics.yaml:1624]: I've added logging for both errors and user cancellations. Thanks for the nudge, I agree it's much stronger with that.

It took a bit, because we were initially logging errors with every user cancellation OR missing some errors, if we logged errors in a different spot. 

I believe the "logging errors with each cancellation" was happening because we were incorrectly considering a user-cancelled DNS lookup an error. (A user would cancel the speed check, we'd intentionally cancel the upload and download tasks, but a DNS error of user-cancellation would still trigger the code to set an error state.) I believe I've cleared this up in `benchmarktasktransfer.cpp`.

I believe I'm currently only logging one end state for the end of any speed test, but since there are many ways they could end, I didn't see a single place in the code to handle all cases, unless I added a new Cancelled state to ConnectionBenchmark, which seemed heavy handed and would bring risks.

File: glean/metrics.yaml
Comment (brizental) [reply glean/metrics.yaml:1733]: Yes I agree about not requiring new Glean.

File: tests/qml/tst_GleanEvents.qml
Comment (brizental) [inline tests/qml/tst_GleanEvents.qml:1]: Let's just remove these tests for now and work on proper tests once we have looked into how to do that :)

File: glean/metrics.yaml
Comment (brizental) [reply glean/metrics.yaml:1670]: > Santiago was most concerned with "what error does the user see", and seemingly less concerned about what is happening technically behind the scenes.

I would argue that you can infer what the user is seeing from what is happening behind the scenes. But I won't block this anymore.

**Review by brizental — APPROVED** [review]

File: tests/qml/tst_GleanEvents.qml
Comment (mcleinman) [reply tests/qml/tst_GleanEvents.qml]: Removed! I believe that's the last outstanding conversation - let me know if you want any other tweaks.

File: src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp
Comment (mcleinman) [reply src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp]: This has now been updated, and we're now tracking errors and cancellations.

Comment (mcleinman) [reply src/apps/vpn/connectionbenchmark/connectionbenchmark.cpp]: @bakulf  I'm going to resolve this and merge on Monday, unless you speak up and ask to do another review. (I'd love to have another review if you want, but Bea has also spent a bunch of time on this one - very much up to you.)

**Review by bakulf — APPROVED** [review]
r+

**Pull request merged**
